app/services/teacher_training_public_api/sync_sites.rb

```diff
@@ -49,91 +62,42 @@ def sync_site(site_from_api)
-      (from_existing_course_options + [course.study_mode]).uniq
-    end
-
-    def create_course_options(site, study_mode, _site_status)
+    def create_or_update_course_option(site, study_mode)
       course_option = CourseOption.find_or_initialize_by(
-        site:,
         course_id: course.id,
+        site:,
         study_mode:,
       )
 
-      course_option.update!(vacancy_status:)
+      course_option.update!({
+        site_still_valid: true,
+        vacancy_status: vacancies_for(course, study_mode),
+      })
 
       @updates.merge!(course_option: true) if !@incremental_sync
     end
 
-    def vacancy_status
-      case @course_status_from_api
-      when 'open'
-        'vacancies'
-      when 'closed'
-        'no_vacancies'
-      else
-        raise InvalidVacancyStatusDescriptionError, @course_status_from_api
+    def disable_or_delete_obsolete_course_options(course, api_site_uuids)
+      course_options_for_deletion = course.reload.course_options.select do |course_option|
+        !course_option.study_mode.in?(course.study_modes) || !course_option.site.uuid.in?(api_site_uuids)
       end
-    end
-
-    class InvalidVacancyStatusDescriptionError < StandardError; end
-
-    def handle_course_options_with_invalid_sites(sites)
-      course_options = @course.course_options.joins(:site)
-      site_uuids = sites.map(&:uuid)
-      invalid_course_options = course_options.where.not(site: { uuid: site_uuids })
-      return if invalid_course_options.blank?
-
-      chosen_course_option_ids = ApplicationChoice
-                                     .where(course_option: invalid_course_options)
-                                     .or(ApplicationChoice.where(current_course_option: invalid_course_options))
-                                     .pluck(:course_option_id, :current_course_option_id).flatten.uniq
 
-      not_part_of_an_application = invalid_course_options.where.not(id: chosen_course_option_ids)
-      not_part_of_an_application.delete_all
-      part_of_an_application = invalid_course_options.where(id: chosen_course_option_ids)
-
-      return if part_of_an_application.empty?
-
-      part_of_an_application.each do |course_option|
-        next if course_option.site_still_valid == false
-
-        course_option.update!(site_still_valid: false)
+      course_options_for_deletion.each do |course_option|
+        if course_option.application_choices.any? || course_option.current_application_choices.any?
+          course_option.update(site_still_valid: false, vacancy_status: :no_vacancies)
+        else
+          course_option.destroy
+        end
       end
     end
 
-    def handle_course_options_with_reinstated_sites(sites)
-      withdrawn_course_options = @course.course_options.joins(:site).where(site_still_valid: false)
-      site_uuids = sites.map(&:uuid)
+    def vacancies_for(course, study_mode)
+      return :no_vacancies if @course_status_from_api == 'closed'
```

Comment:
Is it possible to adapt the method below to handle `study_mode`? It might not be worth it but just asking.

https://github.com/DFE-Digital/apply-for-teacher-training/blob/c8977340dd5dc3a85c07ae0f31d056f2c0fdd0c8/app/services/teacher_training_public_api/sync_sites.rb#L90-L99

Reply:
Yep, i'm happy with your implementation, it just felt like two very similar methods.